ingester2/src/wal/reference_tracker.rs

```diff
@@ -0,0 +1,646 @@
+    /// The returned [`WalReferenceActor`] SHOULD be
+    /// [`WalReferenceActor::run()`] before the handle is used to avoid
+    /// potential deadlocks.
+    pub(crate) fn new<T>(wal: T) -> (Self, WalReferenceActor<T>)
+    where
+        T: WalFileDeleter,
+    {
+        let (file_tx, file_rx) = mpsc::channel(5);
+        let (persist_tx, persist_rx) = mpsc::channel(50);
+        let (unbuffered_tx, unbuffered_rx) = mpsc::channel(50);
+
+        let actor = WalReferenceActor {
+            wal,
+            persisted: SequenceNumberSet::default(),
+            wal_files: HashMap::with_capacity(3),
+            file_rx,
+            persist_rx,
+            unbuffered_rx,
+        };
+
+        (
+            Self {
+                file_tx,
+                persist_tx,
+                unbuffered_tx,
+            },
+            actor,
+        )
+    }
+
+    /// Enqueue a new file rotation event, providing the [`SegmentId`] of the
+    /// WAL file and the [`SequenceNumberSet`] the WAL segment contains.
+    pub(crate) async fn enqueue_rotated_file(&self, segment_id: SegmentId, set: SequenceNumberSet) {
+        Self::send(&self.file_tx, (segment_id, set)).await
+    }
+
+    /// Enqueue a persist completion notification for newly persisted data.
+    pub(crate) async fn enqueue_persist_notification(&self, note: Arc<CompletedPersist>) {
+        Self::send(&self.persist_tx, note).await
+    }
+
+    /// Enqueue a notification that a write appearing in some WAL segment will
+    /// not be buffered/persisted (either the active, not-yet-rotated segment or
+    /// a prior, already-rotated segment).
+    ///
+    /// This can happen when a write is added to the WAL segment and
+    /// subsequently fails to be applied to the in-memory buffer. It is
+    /// important to track these unusual cases to ensure the WAL file is not
+    /// kept forever due to an outstanding reference, waiting for the unbuffered
+    /// write to be persisted (which it never will).
+    pub(crate) async fn enqueue_unbuffered_write(&self, id: SequenceNumber) {
+        Self::send(&self.unbuffered_tx, id).await
+    }
+
+    /// Send `val` over `chan`, logging a warning if `chan` is at capacity.
+    async fn send<T>(chan: &mpsc::Sender<T>, val: T)
+    where
+        T: Debug + Send,
+    {
+        match chan.try_send(val) {
+            Ok(()) => {}
+            Err(TrySendError::Full(val)) => {
+                warn!(?val, "notification buffer is full");
+                chan.send(val).await.expect("wal reference actor stopped");
+            }
+            Err(TrySendError::Closed(_)) => panic!("wal reference actor stopped"),
+        }
+    }
+}
+
+/// A WAL file reference-count tracker.
+///
+/// See [`WalReferenceHandle`].
+#[derive(Debug)]
+pub(crate) struct WalReferenceActor<T = Arc<wal::Wal>> {
+    wal: T,
+
+    /// The set of IDs of persisted data that do not yet appear in
+    /// `wal_segments`, the set of WAL files rotated out of active use. This is
```

Comment:
```suggestion
    /// `wal_files`, the set of WAL files rotated out of active use. This is
```
is this meant to be the variable you refer to here?